torch/csrc/jit/passes/concat_opt.cpp

```diff
@@ -0,0 +1,480 @@
+#include <torch/csrc/jit/passes/concat_opt.h>
+
+#include <algorithm>
+#include <unordered_set>
+#include <vector>
+
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/ir/named_value.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/constant_pooling.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+void removeCatNodeFromGraph(Node* n) {
+  TORCH_INTERNAL_ASSERT(n->kind() == aten::cat);
+  auto inp_list = n->input(0);
+  GRAPH_UPDATE("Deleting\n", *n);
+  n->destroy();
+  if (!inp_list->hasUses()) {
+    GRAPH_UPDATE("Deleting\n", *inp_list->node());
+    inp_list->node()->destroy();
+  }
+}
+
+class ConcatCseEliminator {
+ public:
+  explicit ConcatCseEliminator(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  void run() {
+    for (auto node : graph_->block()->nodes()) {
+      if (node->kind() == aten::cat) {
+        handleCat(node);
+      }
+    }
+    postprocess();
+  }
+
+ private:
+  void handleCat(Node* node) {
+    GRAPH_DEBUG("Considering cat node for CSE opt: ", node);
+
+    // cat ops are of the following form:
+    //   %inputs = prim::ListConstruct(%inp1, %inp2, %inp3)
+    //   %concat = aten::cat(%inputs, %concat_dim)
+
+    // First, we check if %inputs is same as the inputs of any of the
+    // previous cat ops.
+    //
+    // Example:
+    //     %10 = prim::ListConstruct(%0, %1)
+    //     %11 = aten::cat(%10, ...)
+    //     ...
+    //     %12 = prim::ListConstruct(%0, %1)  // Same as %10
+    //     %13 = aten::cat(%12, ...)
+    //     ...
+    //         = %13 ... // Use %13
+    //
+    // After CSE opt:
+    //     %10 = prim::ListConstruct(%0, %1)
+    //     %11 = aten::cat(%10, ...)
+    //     ...
+    //         = %11 ... // Use %11 in place of %13
+
+    auto curr_list = node->input(0)->node();
+    std::vector<Value*> curr_list_inputs(
+        curr_list->inputs().begin(), curr_list->inputs().end());
+    for (const auto& it : concated_lists_) {
+      auto prev_list = it.first;
+      std::vector<Value*> prev_list_inputs(
+          prev_list->inputs().begin(), prev_list->inputs().end());
+      std::vector<Value*> diff;
+      std::set_difference(
+          curr_list_inputs.begin(),
+          curr_list_inputs.end(),
+          prev_list_inputs.begin(),
+          prev_list_inputs.end(),
+          std::inserter(diff, diff.begin()));
+      if (diff.size() == 0) {
+        redundant_cats_[node] = it.second;
```

Comment:
What if curr_list and prev_list contain the same elements but in different orders, e.g., `aten::cat(%1, %2, %3)` vs. `aten::cat(%2, %1, %3)`? Would `diff.size() == 0` in this case?

Reply:
Probably need to check if mutation happened to any of the elements in between the two concat lists. For instance,

```
%100 = aten::cat(%1, %2, %3)
aten::add_(%1, %1, ...) // in-place update on %1
%200 = aten::cat(%1, %2, %3)
```
In this case, %100 and %200 are not CSE

Reply:
> What if curr_list and prev_list contain the same elements but in different orders, e.g., aten::cat(%1, %2, %3) vs. aten::cat(%2, %1, %3)? Would diff.size() == 0 in this case?

No. `diff` wouldn't be empty in this case. These would be treated as different since their concatenation order is different.

> Probably need to check if mutation happened to any of the elements in between the two concat lists. 

Good point. Added this check and a test for this. I wasn't doing this earlier since I was running it only on the traced graph. But this guard should be there as you pointed out.

Currently, the optimization will not happen for a `cat` op whenever the list that is input to that op has any mutations in the graph (only checking for `aten::append` for now). This could be improved by looking for mutations only in the region of optimization. That is a TODO for later.